### ISSUE TITLE
Should we use multiple packages for rocm-opencl-runtime deps?

### ISSUE THREAD

**Issue by Tzigamm** [emaimx/Solus-ROCm-build#6]
The point of packaged dependencies is to be able to package them once and use them for every package that depends on it, if only one package (rocm-opencl-runtime) depends on them maybe we should make a single tarball hosted somewhere with all the dependencies pre-packed or make rocm-opencl-runtime download everything through its source list (I created a patch for solbuild/ypkg to make this work but it hasn't been merged yet).

When I talked with DataDrake on IRC he was happier with the hosted tarball option, although I never talked about multiple dependencies with him, maybe he would prefer that to other options

**Comment by saitam757** [comment]
If I understand you correctly your idea is to have only one package.yml for the entire rocm-opencl-runtime ? The tarball contains the other tarballs ? The multiple source list seems to be this question I asked sometime ago in the [solus forum](https://discuss.getsol.us/d/1412-dealing-with-more-than-one-source-in-package-yml) ? At that time I looked into arch and gentoo packaging of the Rocm stack and saw that their packaging system allows several sources in one packaging script.
This multiple source list would be the most flexible way in my opinion but in the end the Solus core team has to decide this (and I hope they will merge your patch :-) )

**Comment by Tzigamm** [comment]
Sorry if my message is a bit confusing, I'm french and I wrote that a bit too late for my own good

Yep, that's exactly the problem I had when working on rocm-opencl-runtime. The patch I made handles it by downloading the first one as rocm-3.1.tar.gz, the second one as rocm-3.1.tar.gz.1 and so on.

Since DataDrake plans on completely rewriting solbuild/ypkg, he's not really fond of new features, especially if it's only needed for one package, he suggested we create a script that package everything as one tarball and that he'd be happy to store this tarball on Solus' server

**Comment by emaimx** [comment]
You can already use multiple sources for one package (there is two sources for rocm-opencl-runtime).  But most of the things here are to build rocm and not run it, so at least it make sense to have them separated.

**Comment by Tzigamm** [comment]
Sure, for testing it's alright, but maybe Solus' core team would prefer another approach for when we want to push it to the repo, I'll drop by on IRC and ask them if you want

**Comment by emaimx** [comment]
That is right, if everything is working correctly we might want to have one package with all the build deps and one with all the run deps. However I am not sure how to do that since we actually need to install some of the files to build the rest, so it might be messy.

**Comment by Tzigamm** [comment]
We don't actually need to install anything, here's how I did it using the patch I wrote that allows for multiple files having the same name. DataDrake's idea was to create a tarball where everything is pre-downloaded and pre-linked so we could replace all the links by just this tarball and remove everything in the setup part until the `%patch`

`
name       : rocm-opencl-runtime
version    : 2.10.0
release    : 1
extract    : no
source     :
    - https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime/archive/roc-2.10.0.tar.gz : 7b38b1be2217d7b0ac552e4cfb5fe38660fa59ef61f2a46946bac6fba6efdf7c
    - https://github.com/RadeonOpenCompute/llvm/archive/roc-ocl-2.10.0.tar.gz : c46ec3568b754a7aced6139bcc254d39ddb89016b282c467ae1b5536d8021540
    - https://github.com/RadeonOpenCompute/clang/archive/roc-2.10.0.tar.gz : acdebf229a2120ad0ccdc48a6b0e863a6a4d89f46bb62e7296f2f16bbc4a04a7
    - https://github.com/RadeonOpenCompute/lld/archive/roc-ocl-2.10.0.tar.gz : 105cedd4a31a6d6ad51043c384129a7705f9a5f65414e60a07c9193411b756f0
    - https://github.com/RadeonOpenCompute/ROCm-Device-Libs/archive/roc-ocl-2.10.0.tar.gz : ed56c5f742bd8b2fc7e1124434fa575f51573900444adc235ce6785f3eb56d9d
    - https://github.com/RadeonOpenCompute/ROCm-OpenCL-Driver/archive/roc-2.10.0.tar.gz : 00a50696732a4cd3b880cfb11a64d02c5dfd956a79796eacda166173c14eb97a
    - git|https://github.com/KhronosGroup/OpenCL-ICD-Loader : 978b4b3a29a3aebc86ce9315d5c5963e88722d03
license    : MIT
summary    : ROCm OpenOpenCL Runtime
component  : xorg.display
description: |
    The ROCm OpenCL Runtime
builddeps  :
    - rocr-runtime-devel
    - roct-thunk-interface-devel
    - ocl-icd-devel
    - rocm-cmake
    - mesalib-devel
rundeps    :
    - roct-thunk-interface
    - ocl-icd
setup      : |
    # extract sources
    for F in $sources/*.tar*; do
        tar xf "$F" --directory $workdir/../
    done

    # recreate the file structure expected by the project
    cd ..
    ln -s %workdir%/../ROCm-OpenCL-Runtime-roc-%version% %workdir%/../opencl
    mkdir -p %workdir%/../opencl/compiler
    ln -s %workdir%/../llvm-roc-ocl-%version%/ %workdir%/../opencl/compiler/llvm
    mkdir -p %workdir%/../opencl/compiler/llvm/tools
    ln -s %workdir%/../clang-roc-%version%/ %workdir%/../opencl/compiler/llvm/tools/clang
    ln -s %workdir%/../lld-roc-ocl-%version%/ %workdir%/../opencl/compiler/llvm/tools/lld
    mkdir -p %workdir%/../opencl/library
    ln -s %workdir%/../ROCm-Device-Libs-roc-ocl-%version%/ %workdir%/../opencl/library/amdgcn
    ln -s %workdir%/../ROCm-OpenCL-Driver-roc-%version%/ %workdir%/../opencl/compiler/driver
    mkdir -p $workdir/../opencl/api/opencl/khronos
    ln -s $sources/OpenCL-ICD-Loader.git/ %workdir%/../opencl/api/opencl/khronos/icd

    cd opencl/

    # add a dependency
    # see https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime/issues/64#issuecomment-473502685
    %patch -p1 < $pkgfiles/fix_rocm_opencl_build_order.patch

    sed -i 's:lib/x86_64:lib64/x86_64:' CMakeLists.txt
    sed -i 's:/x86_64:/rocm-opencl:' CMakeLists.txt

    %cmake_ninja ..
build      : |
    cd $workdir/../opencl/
    %ninja_build

    # has to be built manually since 2.3.0
    %ninja_build lld
install    : |
    cd $workdir/../opencl/
    %ninja_install

    mkdir -p $installdir/usr/share/OpenCL/vendors/
    echo "/usr/lib64/rocm-opencl/libamdocl64.so" > $installdir/usr/share/OpenCL/vendors/rocm-opencl64.icd

    install -dm644 $installdir/usr/share/ld.so.conf.d/
    cp $pkgfiles/rocm-opencl.conf $installdir/usr/share/ld.so.conf.d/rocm-opencl.conf

    rm -f $installdir/usr/lib64/libOpenCL.so*
`
